test/standalone/csv_test.cpp

```diff
@@ -0,0 +1,715 @@
+    }
+    CHECK(threw);
+  } // END SECTION
+
+  SECTION("feature ID only incremented for valid rows") {
+    auto ds = get_csv_ds("test/data/csv/warns/feature_id_counting.csv", false);
+    auto fs = all_features(ds);
+
+    // first
+    auto feature = fs->next();
+    REQUIRE(bool(feature));
+    CHECK(feature->id() == 1);
+
+    // second, should have skipped bogus one
+    feature = fs->next();
+    REQUIRE(bool(feature));
+    CHECK(feature->id() == 2);
+
+    feature = fs->next();
+    CHECK(!feature);
+  } // END SECTION
+
+  SECTION("dynamically defining headers") {
+    using ustring = mapnik::value_unicode_string;
+    using row = std::pair<std::string, std::size_t>;
+
+    for (auto const &r : {
+          row{"test/data/csv/fails/needs_headers_two_lines.csv", 2}
+        , row{"test/data/csv/fails/needs_headers_one_line.csv", 1}
+        , row{"test/data/csv/fails/needs_headers_one_line_no_newline.csv", 1}
+      }) {
+      mapnik::parameters params;
+      params["type"] = std::string("csv");
+      params["file"] = r.first;
+      params["headers"] = "x,y,name";
+      auto ds = mapnik::datasource_cache::instance().create(params);
+      REQUIRE(bool(ds));
+
+      auto fields = ds->get_descriptor().get_descriptors();
+      require_field_names(fields, {"x", "y", "name"});
+      require_field_types(fields, {mapnik::Integer, mapnik::Integer, mapnik::String});
+      require_attributes(all_features(ds)->next(), {
+          attr{"x", 0}, attr{"y", 0}, attr{"name", ustring("data_name")} });
+      REQUIRE(count_features(all_features(ds)) == r.second);
+    }
+  } // END SECTION
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wlong-long"
+  SECTION("64bit int fields work") {
+    auto ds = get_csv_ds("test/data/csv/64bit_int.csv");
+    auto fields = ds->get_descriptor().get_descriptors();
+    require_field_names(fields, {"x", "y", "bigint"});
+    require_field_types(fields, {mapnik::Integer, mapnik::Integer, mapnik::Integer});
+
+    auto fs = all_features(ds);
+    auto feature = fs->next();
+    require_attributes(feature, {
+        attr{"x", 0}, attr{"y", 0}, attr{"bigint", 2147483648} });
+
+    feature = fs->next();
+    require_attributes(feature, {
+        attr{"x", 0}, attr{"y", 0}, attr{"bigint", 9223372036854775807ll} });
+    require_attributes(feature, {
+        attr{"x", 0}, attr{"y", 0}, attr{"bigint", 0x7FFFFFFFFFFFFFFFll} });
+  } // END SECTION
+#pragma GCC diagnostic pop
+
+  SECTION("various number types") {
+    auto ds = get_csv_ds("test/data/csv/number_types.csv");
+    auto fields = ds->get_descriptor().get_descriptors();
+    require_field_names(fields, {"x", "y", "floats"});
+    require_field_types(fields, {mapnik::Integer, mapnik::Integer, mapnik::Double});
+
+    auto fs = all_features(ds);
+    for (double d : { .0, +.0, 1e-06, -1e-06, 0.000001, 1.234e+16, 1.234e+16 }) {
+      auto feature = fs->next();
+      REQUIRE(bool(feature));
+      CHECK(feature->get("floats") == d);
```

Comment:
When comparing floats its typically a good idea to use 'Approx' which is also part of catch.hpp. An example of it being used is [here](https://github.com/mapnik/mapnik/blob/master/test/unit/geometry/geometry_equal.hpp#L121).